SOLR-16641 - Generate gradle.properties from gradlew

colvinco opened the pull request:
https://issues.apache.org/jira/browse/SOLR-16641

<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Solr:

* https://issues.apache.org/jira/projects/SOLR

For something minor (i.e. that wouldn't be worth putting in release notes), you can skip JIRA. 
To create a Jira issue, you will need to create an account there first.

The title of the PR should reference the Jira issue number in the form:

* SOLR-####: <short description of problem or changes>

SOLR must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

Stop generating gradle.properties and use a checked-in template instead. If `gradle.properties` isn't present when gradlew runs, it will be created from the `gradle.properties.template`. The properties will be available for the first execution of gradle on a clean checkout.

Also added notes on the need to have `perl` and `python3` for `gradlew check` to succeed.

# Solution

The gradlew already had hooks that altered the arguments passed in when gradle.properties didn't exist. I've changed them to create the gradle.properties instead, and removed the `localSettings` task from gradle.

# Tests

Without this PR
1. Delete gradle.properties
2. Run `./gradlew updateLicenses` and it will always result in a circular dependency and fail.
3. Apply the PR and run `./gradlew updateLicenses` again and it will generate the gradle.properties and succeed.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

uschindler commented:
Hi @dweiss,
I am not fully sure if this is a good idea to do it like that.
In reality there should be an option in gradle to apply settings at runtime (e.g., using the global build script which sets up project structure) and allow to load custom/default settings and/or our own settings.

magibney commented:
There was some related discussion recently on [SOLR-16634](https://issues.apache.org/jira/browse/SOLR-16634?focusedCommentId=17680614#comment-17680614) -- at the moment I'm inclined to favor @dsmiley's suggestion to possibly move localSettings into the `gradlew` wrapper script.

dweiss commented on gradle/validation/jar-checks.gradle:
```diff
@@ -313,7 +313,9 @@ configure(plist) {
       .flatten()
       .each { task ->
-        task.mustRunAfter updateLicenses
+        if (task.name != 'localSettings') {
```
Perhaps use the full path here?

colvinco commented:
I've moved the generation of the localSettings to gradlew as discussed on the Jira issue.
I've also removed the task dependencies between localSettings and other tasks as they're redundant when executed through gradlew.
I've also removed the CI steps that called localSettings

risdenk commented on gradlew:
```diff
@@ -163,11 +163,8 @@ fi
-if [ ! -e "$APP_HOME/gradle.properties" ]; then
-    GRADLE_DAEMON_CTRL=--no-daemon
-fi
+# Hold on to the original arguments, don't want to pass them in when generating gradle.properties
+TASK_ARGS=$@
```
Not sure I understand this line with what is on lines 273 and 282 - it looks like "$@" is still used in both places? Also it doesn't look like we do the same TASK_ARGS trick for Windows if its needed?

Also does $@ need to be quoted?

colvinco replied:
I'm not going to claim to be an expert at bash, so I could be wrong on the details of this. I have tested it though.

The positional parameters referenced by `$@` are mutated by the `set` cmd at various points in the script 
https://github.com/apache/solr/blob/90acc20ef22680966e6708f8985f7e110e6a7700/gradlew#L226
https://github.com/apache/solr/blob/90acc20ef22680966e6708f8985f7e110e6a7700/gradlew#L236
https://github.com/apache/solr/blob/90acc20ef22680966e6708f8985f7e110e6a7700/gradlew#L268
which push more arguments onto the parameters. That's described in https://github.com/apache/solr/blob/90acc20ef22680966e6708f8985f7e110e6a7700/gradlew#L45-L48

Initially `$@` will refer to the parameters passed in when execution of `gradlew` began, i.e. any parameters supplied by the user `./gradlew updateLicenses check -x test --dry-run` -> `updateLicenses check -x test --dry-run`. The set commands are pushing additional arguments into that, for the JRE, for the main class `org.gradle.wrapper.GradleWrapperMain`, etc.

So when https://github.com/apache/solr/blob/90acc20ef22680966e6708f8985f7e110e6a7700/gradlew#L275 is called it contains all the arguments for executing gradle, including the user supplied arguments. However, for calling `localSettings` I don't want the user supplied arguments, so the easiest way I could see to do it without rewriting the whole script was to stash them and not append them into the arguments being built with `set`.

The Windows .bat file doesn't do any complex argument building, so I didn't need to mess around in it to get the same effect.

> Also does $@ need to be quoted?
Quoting it here appeared to quote the entire string of arguments rather than quoting the individual arguments. Maybe I'm missing a trick as that sounds like it would be the normal expansion behavior https://stackoverflow.com/a/3898681

dweiss commented:
> In reality there should be an option in gradle to apply settings at runtime (e.g., using the global build script which sets up project structure) and allow to load custom/default settings and/or our own settings.

In reality, gradle folks are not really responsive to such requests... I know. I wish it was an option too.

magibney commented:
btw I wasn't intending to throw the scope of this PR by mentioning extracting `gradle.properties` generation into the gradlew script. I think it's still an open question whether it makes sense to pursue that. But it definitely makes sense to fix the circular dependency (as you did in aabacb56472acb573ad60b8b5d21913e3a8d0af5).

Maybe it'd make sense to proceed with a different PR for the gradlew script quesiton? (sorry for the mixed messages).

colvinco commented:
np. I don't mind doing a separate PR if it makes it easier to discuss and get the https://github.com/apache/solr/commit/aabacb56472acb573ad60b8b5d21913e3a8d0af5 fix back in this one, or just let this one brew.
I don't think there's a huge rush to fix this so happy to do it either way

dsmiley reviewed (commented):
Yay; glad to see you working on this @colvinco !  It's a big deal!  It'll improve the user getting-started experience (and even for experienced devs that forget localSettings LOL) and improve CI maintenance (e.g. working on Jenkins, GitHub Actions, crave.io).  I think this PR/JIRA issue should more obviously declare this.  It's a 100x more important than a "circular task dependency" in the build.

I'm surprised to see the "localSettings" gradle task live on.  Surely it could be replaced with a bash script?

dweiss commented:
> I'm surprised to see the "localSettings" gradle task live on. Surely it could be replaced with a bash script?

Depends. Some values there are computed from Java defaults. I agree a modification to launch scripts is probably easier to work with though, especially because it can pick up those defaults on the first run.

colvinco commented:
It's also having to "write it twice" for the Windows .bat as long as it's still supported for dev. (Though with WSL becoming more prevalent that's a problem that will solve itself eventually).

colvinco commented:
> A simplifying alternative would be to check gradle.properties into the project (but still marked ignored in .gitignore) and with Gradle's own defaults for number of workers & tests, or choose low ones. Advise users to edit them according to their machine. We are overthinking this stuff today! This has dev maintenance costs we pay right now. Generating this file is over-engineered IMO.

I would create a `gradle.properties.template` file that lives in source control and copy it to the ignored `gradle.properties` path. That way it's easy to update the template without messing around with gitignore.

I've taken a closer look at the properties file. Currently I see that only two of the settings are generated dynamically, and the rest are all hardcoded.
```
        // Approximate a common-sense default for running gradle/tests with parallel
        // workers: half the count of available cpus but not more than 12.
        def cpus = Runtime.runtime.availableProcessors()
        def maxWorkers = (int) Math.max(1d, Math.min(cpus * 0.5d, 12))
        def testsJvms = (int) Math.max(1d, Math.min(cpus * 0.5d, 12))
...
# Maximum number of parallel gradle workers.
org.gradle.workers.max=${maxWorkers}

# Maximum number of test JVMs forked per test task.
tests.jvms=${testsJvms}
```
Getting the number of "processors" (depending on whether hyperthreading counts...) shouldn't be a problem on bash, https://stackoverflow.com/questions/6481005/how-to-obtain-the-number-of-cpus-cores-in-linux-from-the-command-line, so I could still seed in those values.

I'm using Ubuntu in WSL:
```
me:~$ getconf _NPROCESSORS_ONLN
16
me:~$ nproc
16
me:~$ grep -c ^processor /proc/cpuinfo
16
me:~$ grep ^cpu\\scores /proc/cpuinfo | uniq |  awk '{print $4}'
8
```

For Windows it's also possible: https://pureinfotech.com/check-many-cores-processor-windows-10/
```
wmic cpu get NumberOfCores,NumberOfLogicalProcessors /format:value
NumberOfCores=8
NumberOfLogicalProcessors=16
```

`Runtime.runtime.availableProcessors()` gives the same answer as `nproc`, but it was halved (and capped to 12). So I could just use the number of cores, or divide `nproc`.

Doing the string replacement when copying the file in bash shouldn't be a problem, I'll see if I can do it sensibly in batch as well.

colvinco commented on gradle.properties.template:
```diff
@@ -0,0 +1,43 @@
+# timeout after 15 mins of inactivity.
+org.gradle.daemon.idletimeout=900000
+
+# Maximum number of parallel gradle workers.
+org.gradle.workers.max=8
+
+# Maximum number of test JVMs forked per test task.
+tests.jvms=8
```
Can set some different defaults, but they will only affect Windows, as they will get overridden on Linux.

uschindler replied:
Please use a low value like "2" as default (also for max workers). If I start gradle on my laptop i dont want the machine to crush.

colvinco commented:
Hopefully we're getting there now :)
I didn't bother with trying to do variable substitution from batch as I didn't see a good + easy option. But bash will set the two parameters that were being set before.

colvinco commented:
The check on crave.io is failing to apply my commit... I'm guessing that's because I've been force pushing to my branch and it's now a bit out of sync? https://github.com/apache/solr/actions/runs/4063490165/jobs/6995782967

```
error: patch failed: gradlew.bat:80
error: gradlew.bat: patch does not apply
hint: Use 'git am --show-current-patch=diff' to see the failed patch
Applying: SOLR-16632: Add core name to periodic delete related log messages (#1311)
Applying: SOLR-16641 - Generate gradle.properties from template
Patch failed at 0002 SOLR-16641 - Generate gradle.properties from template
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```

I can create a clean PR on a fresh branch to test that, unless there's another reason for it to happen?

uschindler commented on gradle/generation/local-settings.gradle:
```diff
@@ -45,83 +45,4 @@ configure(rootProject) {
       }
     }
   }
-
```
Please keep the "deprecated" localSettings task fora while as all jenkins/github jobs execute them. We can remove it later, but for now add a "fake" task for compatibility.

risdenk replied:
@colvinco I think this is still outstanding

uschindler reviewed (changes requested):
I'd also get this at some point into Lucene's build.

uschindler commented:
> The check on crave.io is failing to apply my commit... I'm guessing that's because I've been force pushing to my branch and it's now a bit out of sync? https://github.com/apache/solr/actions/runs/4063490165/jobs/6995782967
> 
> ```
> error: patch failed: gradlew.bat:80
> error: gradlew.bat: patch does not apply
> hint: Use 'git am --show-current-patch=diff' to see the failed patch
> Applying: SOLR-16632: Add core name to periodic delete related log messages (#1311)
> Applying: SOLR-16641 - Generate gradle.properties from template
> Patch failed at 0002 SOLR-16641 - Generate gradle.properties from template
> When you have resolved this problem, run "git am --continue".
> If you prefer to skip this patch, run "git am --skip" instead.
> To restore the original branch and stop patching, run "git am --abort".
> ```
> 
> I can create a clean PR on a fresh branch to test that, unless there's another reason for it to happen?

Don't force-push in PRs :-)

PRs are anyways squash merged, so no need to squash them before. It also makes it hard to review when you do changes.

colvinco commented:
> Don't force-push in PRs :-)
> 
> PRs are anyways squash merged, so no need to squash them before. It also makes it hard to review when you do changes.

Yep, I normally I wouldn't. But because the solution changed so drastically it seemed better this time around.
Good to hear that squash merge is now the default/only option on here. I had a previous PR that got merged in with a standard merge last year, which was a bit messy :)

colvinco commented on gradlew:
```diff
@@ -163,10 +163,18 @@ fi
+    echo "Populating gradle.properties from gradle.properties.template."
+    # Do the copy first. If the next steps fail, the gradle.properties will still be usable
+    cp gradle.properties.template gradle.properties
+    NUM_CORES=$(grep ^cpu\\scores /proc/cpuinfo | uniq |  awk '{print $4}')
```
I've had another look. I think I can make it work for windows as well without it being too horrible.

janhoy replied:
Also, won't work on macOS :(

dweiss replied:
I really don't want to stand in the way here but I think a java-file based solution (like WrapperDownloader) could prove to be a good middle ground between the needs to support different platforms (windows, mac, linux) and to speed up that initial gradle.properties defaults.

uschindler replied:
aren't we not executing a small java script anyways from Gradlew? I think it mainly downloads gradle distribution.

My idea would be to modify this script/class (executed without compiling directly with java) to write the properties file.

colvinco replied:
I must admit, I hadn't realized Java 11+ supported execution without compilation, I didn't pick up on that being used by gradlew. That's cool (even if it is limited to self contained source files)

I will switch to that, that's much better.

> My idea would be to modify this script/class (executed without compiling directly with java) to write the properties file.

I see [WrapperDownloader takes an arg](https://github.com/apache/solr/blob/main/buildSrc/src/main/java/org/apache/lucene/gradle/WrapperDownloader.java#L48-L51) for the destination it downloads the gradle wrapper jar to.
Do you think it's preferable to add this functionality to the WrapperDownloader and it just grows a little, or create a separate class and invoke it on its own?

colvinco replied:
For now I've put it into a new class, but I can easily move it in the wrapper downloader

colvinco commented:
Yep I used to do that, but it makes it (slightly) harder to update the defaults in the future and can be confusing because people don't expect gitignored files to be in git - as you say people don't realize that it's an option. It also means that if you copy the source of a repo (e.g. from https://github.com/apache/solr/archive/refs/heads/main.zip), and then create a new repo from it you lose the ignored files.

So that's why I copy templates instead. But I don't mind doing whatever people prefer here.

dweiss commented:
> We can absolutely put a real default safe (albeit slow) gradle.properties here and the existing .gitignore rule that already references gradle.properties will prevent someone from accidentally committing their changes.

I'm not sure what you mean. When you have a file that's versioned and also part of .gitignore, this file is versioned - any changes to it will be indicated in git status, git add, etc. It's also committed as if .gitignore didn't include that file. This rules out any "ignored local modifications", unfortunately.

dweiss commented:
If you're concerned about startup time of that one-time "gradle localSettings" process then I'd suggest implementing this template logic/ retrieval of what's necessary as a java class - much more lightweight compared to full gradle. It's pretty much the same trick that's done with WrapperDownloader.java currently.

janhoy commented on gradlew:
```diff
@@ -163,10 +163,18 @@ fi
+    if [ $NUM_CORES -ge 12 ]; then
+      NUM_CORES=12
+    fi
+    sed -i "s/org.gradle.workers.max=[[:digit:]]\+/org.gradle.workers.max=${NUM_CORES}/g;s/tests.jvms=[[:digit:]]\+/tests.jvms=${NUM_CORES}/g" gradle.properties
```
This modification could be wrapped in a test for whether $NUM_CORES is in fact a positive integer, to avoid replacing default with invalid value.

janhoy commented on gradlew:
```diff
@@ -163,10 +163,18 @@ fi
-    GRADLE_DAEMON_CTRL=--no-daemon
+    echo "Populating gradle.properties from gradle.properties.template."
+    # Do the copy first. If the next steps fail, the gradle.properties will still be usable
+    cp gradle.properties.template gradle.properties
```
```suggestion
    cp gradle.properties.template "$APP_HOME/gradle.properties"
```

Also, could the `gradle.properties.template` file be inside `gradle/` folder to avoid littering the root git folder?

risdenk commented on gradlew:
```diff
@@ -163,11 +163,16 @@ fi
-    GRADLE_DAEMON_CTRL=--no-daemon
+    "$JAVACMD" $JAVA_OPTS --source 11 "$APP_HOME/buildSrc/src/main/java/org/apache/lucene/gradle/GradlePropertiesGenerator.java" "$APP_HOME/gradle/template.gradle.properties" "$APP_HOME/gradle.properties"
+    GENERATOR_STATUS=$?
+    if [ "$WRAPPER_STATUS" -ne 0 ]; then
+        exit $WRAPPER_STATUS
+    fi
```
Should use `GENERATOR_STATUS`?

risdenk commented on gradle/template.gradle.properties:
```diff
@@ -0,0 +1,43 @@
+# See gradlew :helpLocalSettings for more information on configuring this file.
```
Not sure this comment is still valid.

colvinco replied:
That command will still give the help output from https://github.com/apache/solr/blob/main/help/localSettings.txt which describes the concepts, rather than the original gradle task.

It could be moved into the template file as a comment instead though

risdenk replied:
Yea my comment was about `configuring this file` - agree the concepts are good. I'm not saying we need to change this completely, but it would be good to consolidate to avoid having to go look another place.

colvinco replied:
I've moved the help content into the template so it's all self contained

risdenk commented on gradlew:
```diff
@@ -163,11 +163,16 @@ fi
+# Generate gradle.properties if they don't exist
 if [ ! -e "$APP_HOME/gradle.properties" ]; then
-    GRADLE_DAEMON_CTRL=--no-daemon
+    "$JAVACMD" $JAVA_OPTS --source 11 "$APP_HOME/buildSrc/src/main/java/org/apache/lucene/gradle/GradlePropertiesGenerator.java" "$APP_HOME/gradle/template.gradle.properties" "$APP_HOME/gradle.properties"
```
`buildSrc/src/main/java/org/apache/lucene/gradle/GradlePropertiesGenerator.java` I think this file wasn't checked in?

colvinco replied:
The gitignore was ignoring everything lucene/. I've updated it to not ignore buildSrc\src\main\java\org\apache\lucene now

risdenk reviewed (changes requested):
I like the direction this is going even though its been a bit winding to get there.

colvinco commented:
It's been a bit of a rollercoaster ;)

risdenk commented on .gitignore:
```diff
@@ -38,6 +38,7 @@ __pycache__
 
 # Ignore lucene included build
 lucene/
```
so I thought the `lucene/` entry was in case of having local Lucene development? It might be worth changing to `/lucene/` or something similar that matches top level only? 

I think its a bit surprising that `lucene/` matches any folder in the file structure.

risdenk commented on buildSrc/src/main/java/org/apache/lucene/gradle/GradlePropertiesGenerator.java:
```diff
@@ -0,0 +1,96 @@
+                  final String newLine = key + '=' + value;
+                  System.out.println("Setting " + newLine);
+                  return newLine;
+              }
```
nit: indent is off

risdenk commented on buildSrc/src/main/java/org/apache/lucene/gradle/GradlePropertiesGenerator.java:
```diff
@@ -0,0 +1,96 @@
+import java.nio.file.StandardOpenOption;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.*;
```
nit: I'm guessing that spotless doesn't run against these files (thats fine) but usually we avoid wildcard imports.

colvinco replied:
Yep it didn't run on them. I've run it manually for next commit.

risdenk reviewed (approved):
I'm happy with this with some minor cleanup (optional)

colvinco commented on gradle/template.gradle.properties:
```diff
@@ -0,0 +1,106 @@
+
+# Set these to enable automatic JVM location discovery.
+org.gradle.java.installations.fromEnv=JDK11,JDK12,JDK13,JDK14,JDK15,JDK16,JDK17
+org.gradle.java.installations.paths=(custom paths)
```
Actually I suppose this one should be commented out. I assume `(custom paths)` isn't a special value, it's a placeholder?
```suggestion
# org.gradle.java.installations.paths=(custom paths)
```

janhoy replied:
Agree, see https://docs.gradle.org/current/userguide/toolchains.html#sec:custom_loc

dweiss commented on buildSrc/src/main/java/org/apache/lucene/gradle/GradlePropertiesGenerator.java:
```diff
@@ -0,0 +1,86 @@
+    var replacements = Map.of("org.gradle.workers.max", maxWorkers, "tests.jvms", testsJvms);
+
+    // Java properties doesn't preserve comments, so going line by line instead
+    // The properties file isn't big, nor is the map of replacements, so not worrying about speed
```
This will work for simple values (and property names) but not when property values are more complex characters - these should (in theory at least) be escaped and/or encoded to plain ascii. I don't think it's a problem for now.

colvinco replied:
Done

dweiss commented on help/localSettings.txt:
```diff
@@ -1,62 +0,0 @@
-Local developer settings
-========================
-
-The first invocation of any task in Solr gradle build will generate
```
I think there is some value in leaving (some content) of this file under help/ - so that folks know that this file can be tweaked and is generated?

colvinco replied:
And keep it wired into `gradlew :helpLocalSettings`?

dweiss replied:
Sorry - I thought I removed  that comment. This is duplicated in the template now so I think it's fine to be removed.

colvinco replied:
Ah bad timing. I can revert that last commit I just put in, or leave it. It might still be a useful signpost

dweiss reviewed (approved):
LGTM.

uschindler reviewed (changes requested):
See my comment, I don't like the explicit parsing of the properties file. Just see it as a plain text template file and search/replace some ANT-like `@VAR@` patterns in it: Load as string, replace tokens, save string (as UTF-8).

In general I am still not really happy with the whole setup. At moment in Lucene I have see this problem very often: You add some new features (like the MR JAR and adding new Java versions) and then many people have the template file expanded in their checkout and ten build fails, because it is not updated.

Not sure how to solve this without Gradle allowing to set some of the properties at runtime.

I have some ideas, but actually we should really open a bug reort that they finally fix the stupidness with toolkits, those properties files and allowing to set those properties in the gradle.settings file using code (which is committed to repo).

uschindler commented on gradle/template.gradle.properties:
```diff
@@ -0,0 +1,106 @@
+
+# Set these to enable automatic JVM location discovery.
+org.gradle.java.installations.fromEnv=JDK11,JDK12,JDK13,JDK14,JDK15,JDK16,JDK17
+org.gradle.java.installations.paths=(custom paths)
```
This needs to be commented out (see Lucene build). It hurts when you actually want to use toolkits.

janhoy commented on gradle/template.gradle.properties:
```diff
@@ -0,0 +1,108 @@
+org.gradle.java.installations.auto-download=false
+
+# Set these to enable automatic JVM location discovery.
+org.gradle.java.installations.fromEnv=JDK11,JDK12,JDK13,JDK14,JDK15,JDK16,JDK17
```
Is this the JVM that gradle itself runs under? I suppose Solr build itself will run under whatever JVM is active in the shell that starts the build?

uschindler replied:
No thats only needed for Toolkits. It is unused by Solr ant moment. IMHO, this can go away for Solr, because when you set it up you need to change it anyways.

Lucene uses it heavily to find JDK19 and JDK20 to build the MR-JAR.

janhoy replied:
Yea, I was unaware of toolchains and read up a bit. If unused by Solr the lines should just go away until we start using it...

colvinco replied:
Done

janhoy reviewed (approved):
Checked out and tested. Verified that props file was created. On my M1 Pro mac with 10 cores the jvms were set to 5. Looks good!

janhoy commented on buildSrc/src/main/java/org/apache/lucene/gradle/GradlePropertiesGenerator.java:
```diff
@@ -0,0 +1,71 @@
+      throw new IOException("template file not found: " + source);
+    }
+    if (Files.exists(destination)) {
+      System.out.println(destination + " already exists, skipping generation.");
```
This line prints a message that it skips generation, but there is no exit, so it continues generate and overwrite...

colvinco replied:
Woops :)

janhoy replied:
It would never happen, since in `gradlew` you only run this if it does not exist... Anyway...

uschindler reviewed (approved):
Works as first step.
Could you make a PR for Lucene, too (please refer to discussions here).

madrob commented on buildSrc/src/main/java/org/apache/lucene/gradle/GradlePropertiesGenerator.java:
```diff
@@ -0,0 +1,72 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.gradle;
```
Just because this was based on a Lucene class doesn't mean this is a Lucene class. Should probably be packaged under o.a.solr.gradle?

magibney replied:
+1 -- analogous to [this namespace issue](https://github.com/apache/solr/pull/1323/commits/600732f558e3052f5569ab3d189365f9b1edb5d3). Not sure whether there are practical implications to the naming in this case, but regardless, solr package makes sense here.

janhoy replied:
I thought about it too but thought we could leave it for another day. Although it would be natural to change when we're touching this area.

colvinco replied:
I only put it there because it's adjacent to the https://github.com/apache/solr/blob/1d76b053c4a6fb742a97f4250b332d755680d00e/buildSrc/src/main/java/org/apache/lucene/gradle/WrapperDownloader.java that's also being used in gradlew. In actual fact, neither of them belong in buildSrc, because they aren't used by the gradle scripts themselves. I just didn't want to do something different.

I could move them both to https://github.com/apache/solr/tree/1d76b053c4a6fb742a97f4250b332d755680d00e/gradle
Either into a new directory within it, or put the WrapperDownloader in https://github.com/apache/solr/tree/1d76b053c4a6fb742a97f4250b332d755680d00e/gradle/wrapper and the new generator in https://github.com/apache/solr/tree/1d76b053c4a6fb742a97f4250b332d755680d00e/gradle/generation?

uschindler replied:
Remove the package completely. As this is a command line class only and it is not even compiled, theres no need to have a package.

dweiss replied:
I think the reason it was originally placed there was so that it lived together with other "sources" - so that any validation checks, formatting, etc. would apply. I don't think we actually do it to buildSrc (haven't checked) but we could, in which case it'd be easier to maintain it as part of buildSrc than in some ad-hoc location.

janhoy replied:
Can we spin that refactoring out in a separate PR and merge this one, now that it is stable?

uschindler replied:
I would keep the lucene namespace as at moment this makes merging cross source trees easier (I would like to port this over to Lucene, so cherry picking this commit works from there if the path names are similar). At moment I try to port most lucene and solr gradle build imüprovements to the other project, so aligned path names makes this easier.

colvinco replied:
Does someone want to merge it then? :)

janhoy commented:
Please add a line to `CHANGES.txt` for 9.2 release. It can be useful for folks to get alerted about this change.

risdenk commented:
Thanks @colvinco!

uschindler commented:
Thanks, I created https://github.com/apache/lucene/pull/12131 for Apache Lucene (cherrypicking was easy). Please have a look.
